File: src/bon.erl
%% Binary Object Notation
%% A.K.A. Beeno Object Notation
%%
%% Based on stackish.
%%
%% Designed to enable loss-less data exchange between Erlang to Javascript.
%%
%% Loss-less means it will not mess up tuple and list, nor atom and hashmap key
%%
%% integer is represented in dec format, e.g. 42
%% floating number is represented in rational format (of integer), e.g. 314/100

-module(bon).

%% API
-export([encode/1, encode/2, decode/1, decode_all/1]).

%% Util
-export([fac/1]).

%% Testing
-include_lib("eunit/include/eunit.hrl").
-export([data_test/1, fac_test/1, serialize_test/1]).
-export([flatten_test/1]).
-export([debug/0, debug2/0]).

-define(list_to_atom(X), erlang:list_to_existing_atom(X)).

%%%%%%%%%
%% API %%
%%%%%%%%%

encode(Data, Opts) when is_list(Opts) ->
  IOList = serialize(Data),
  case Opts of
    [] ->
      IOList;
    [compact] ->
      shrink(IOList)
  end.

encode(Data) ->
  serialize(Data).

decode(IOList) when is_list(IOList) ->
  decode(iolist_to_binary(IOList));
decode(Bin) when is_binary(Bin) ->
  List = erlang:binary_to_list(Bin),
  {[], Res} = parse(List, []),
  Res.

decode_all(IOList) when is_list(IOList) ->
  decode(iolist_to_binary(IOList));
decode_all(Bin) when is_binary(Bin) ->
  List = erlang:binary_to_list(Bin),
  decode_all(List, []).

%%%%%%%%%%
%% Test %%
%%%%%%%%%%

data_test(Data) ->
  ?assertEqual(Data, decode(encode(Data))).

data_test() ->
  Res = lists:all(fun(X) -> data_test(X) == ok end, [
    42,
    -72,
    3.14,
    atom,
    "AB",
    [65, 0, 66],
    <<"text">>,
    {record, content},
    [{debug, true}, safe, speedup, {log, "file.log"}],
    #{user=>"name", pw=>"123"},
    #{12=>34, "user"=>name},
    #{log=>[#{time=> 123}, #{time=> 234}]},
    sets:from_list([42, 3.14, "str"])
  ]),
  case Res of
    true -> ok;
    X -> X
  end.


%%%%%%%%%%%%%%
%% Internal %%
%%%%%%%%%%%%%%

-spec decode_all(iolist(), [term()]) -> [term()].

decode_all(List, Acc) when is_list(List), is_list(Acc) ->
  case parse(List, []) of
    {[], Res} -> lists:reverse([Res | Acc]);
    {Next_List, Res} -> decode_all(Next_List, [Res | Acc])
  end.

is_string(X) ->
  io_lib:printable_unicode_list(X).

-define(WORD_TUPLE, "t").
-define(WORD_LIST, "l").
-define(WORD_MAP, "m").
-define(WORD_SET, "s").

%%%%%%%%%%%%%%
% serializer %
%%%%%%%%%%%%%%

serialize(X) when is_integer(X) ->
  [$ , erlang:integer_to_binary(X), $ ];
serialize(X) when is_float(X) ->
  Bin = case fac(X) of
          {A, 1} -> erlang:integer_to_binary(A);
          {A, B} -> [erlang:integer_to_binary(A), $/, erlang:integer_to_binary(B)]
        end,
  [$ , Bin, $ ];

serialize(X) when is_atom(X) ->
  Bin = list_to_binary(erlang:atom_to_list(X)),
  Size = erlang:byte_size(Bin),
  Bin_Size = erlang:integer_to_binary(Size),
  [$', $a, $:, Bin_Size, $:, Bin, $'];

serialize(Bin) when is_binary(Bin) ->
  Size = erlang:byte_size(Bin),
  Bin_Size = erlang:integer_to_binary(Size),
  [$', $b, $:, Bin_Size, $:, Bin, $'];

serialize(Tuple) when is_tuple(Tuple) ->
  case sets:is_set(Tuple) of
    true ->
      List = sets:to_list(Tuple),
      Children = serialize_list(List),
      [$[, Children, $ , ?WORD_SET, $ ];
    false ->
      List = erlang:tuple_to_list(Tuple),
      Children = serialize_list(List),
      [$[, Children, $ , ?WORD_TUPLE, $ ]
  end;

serialize(List) when is_list(List) ->
  case is_string(List) of
    true ->
      Str = lists:map(fun quote_string_char/1, List),
      [$", Str, $"];
    false ->
      Children = serialize_list(List),
      [$[, Children, $ , ?WORD_LIST, $ ]
  end;

serialize(Map) when is_map(Map) ->
  Children = maps:fold(
    fun(K, V, Acc) ->
      K_Bin = serialize(K),
      V_Bin = serialize(V),
      [V_Bin, K_Bin | Acc]
    end, [], Map),
  [$[, Children, $ , ?WORD_MAP, $ ].

serialize_list(List) when is_list(List) ->
  lists:foldl(
    fun(X, Acc) ->
      Bin = serialize(X),
      [Bin | Acc]
    end, [], List).

quote_string_char($") -> [$\\, $"];
quote_string_char(C) -> C.

serialize_test(X) ->
  Res =
    binary_to_list(
      iolist_to_binary(
        serialize(X))),
  io:format("~p~n", [Res]).


%%%%%%%%%%
% parser %
%%%%%%%%%%

-record(word, {name}).

-define(is_digit(X), (X >= $0 andalso X =< $9)).
-define(is_small_cap(X), (X >= $a andalso X =< $z)).
-define(is_large_cap(X), (X >= $A andalso X =< $Z)).
-define(is_alphabet(X), (?is_small_cap(X) or ?is_large_cap(X))).
-define(is_word_body(X), (?is_digit(X) or ?is_alphabet(X))).

-spec parse(InputStream, OutputStream) -> {RestInputStream, NewOutputStream} when
  InputStream :: iolist(),
  OutputStream :: [term()],
  RestInputStream :: iolist(),
  NewOutputStream :: [term()].

%% finish
parse([], [Acc]) ->
  {[], Acc};

%% skip spaces
parse([$  | T], Acc) ->
  parse(T, Acc);

%% number
parse([H | T], Acc) when ?is_digit(H) ->
  {Num, Tail} = parse_number(T, H - $0, 1),
  parse(Tail, [Num | Acc]);
parse([$-, H | T], Acc) when ?is_digit(H) ->
  {Num, Tail} = parse_number(T, H - $0, 1),
  parse(Tail, [-Num | Acc]);

%% atom
parse([$', $a, $:, H | T0], Acc) when ?is_digit(H) ->
  {Size, [$: | T1]} = parse_number(T0, H - $0, 2),
  {List, [$' | T2]} = lists:split(Size, T1),
  Atom = ?list_to_atom(List),
  parse(T2, [Atom | Acc]);

%% binary
parse([$', $b, $:, H | T0], Acc) when ?is_digit(H) ->
  {Size, [$: | T1]} = parse_number(T0, H - $0, 2),
  {List, [$' | T2]} = lists:split(Size, T1),
  Bin = list_to_binary(List),
  parse(T2, [Bin | Acc]);

%% group: tuple, list and map
parse([$[ | T0], Acc) ->
  {T1, [#word{name = Name} | Children]} = parse(T0, []),
  Res = case Name of
          ?WORD_TUPLE -> erlang:list_to_tuple(Children);
          ?WORD_LIST -> Children;
          ?WORD_MAP -> list_to_map(Children, #{});
          ?WORD_SET -> sets:from_list(Children)
        end,
  parse(T1, [Res | Acc]);

parse([H | T0], Acc) when ?is_alphabet(H) ->
  {Name, T1} = parse_word(T0, [H]),
  Word = #word{name = Name},
  {T1, [Word | Acc]};

%% string
parse([$" | T0], Acc) ->
  {Str, T1} = parse_string(T0, ""),
  parse(T1, [Str | Acc]).


%%%%%%%%%%%%%%%%%%%%%%
%% Helper Functions %%
%%%%%%%%%%%%%%%%%%%%%%

-spec parse_string(list(), [char()]) -> {[char()], list()}.

parse_string([], Acc) ->
  {lists:reverse(Acc), []};
parse_string([$\\, $" | T], Acc) ->
  parse_string(T, [$" | Acc]);
parse_string([$" | T], Acc) ->
  {lists:reverse(Acc), T};
parse_string([H | T], Acc) ->
  parse_string(T, [H | Acc]).

-spec parse_word(list(), [char()]) -> {[char()], list()}.

parse_word([H | T], Acc) when ?is_word_body(H) ->
  parse_word(T, [H | Acc]);
parse_word(Tail, Acc) when is_list(Tail) ->
  Name = lists:reverse(Acc),
  {Name, Tail}.

-spec parse_number(list(), integer(), 1|2) -> {number(), list()}.

parse_number([H | T], Acc, Count) when ?is_digit(H) ->
  parse_number(T, Acc * 10 + (H - $0), Count);
parse_number([$/, H | T0], Acc, 1) when ?is_digit(H) ->
  {Q, T1} = parse_number(T0, H - $0, 2),
  {Acc / Q, T1};
parse_number(List, Acc, Count) when is_list(List), is_number(Acc), (Count == 1) or (Count == 2) ->
  {Acc, List}.

-spec fac(float()|integer()) -> {integer(), integer()}.

fac(1) -> {1, 1};
fac(F) when F < 1 ->
  {A, B} = fac(1 / F),
  fac(B, A);
fac(F) ->
  {A, B} = fac_power_up(F, 1),
  fac(A, B).

fac_test(X) ->
  {A, B} = fac(X),
  Diff = abs(X - A / B),
  Diff / X.

fac(A, B) ->
  D = gcd(A, B),
  {round(A / D), round(B / D)}.

fac_power_up(F, Acc) when round(F) == F -> {round(F), Acc};
fac_power_up(F, Acc) -> fac_power_up(F * 10, Acc * 10).

gcd(0, X) -> X;
gcd(X, 0) -> X;
gcd(A, B) -> gcd(B, A rem B).

list_to_map([], Acc) when is_map(Acc) ->
  Acc;
list_to_map([K, V | T], Acc) ->
  list_to_map(T, maps:put(K, V, Acc)).

shrink(Xs) when is_list(Xs) ->
  shrink(Xs, []).

shrink([], Acc) ->
  Acc;
shrink([H | T], Acc) when is_list(H) ->
  shrink(H, shrink(T, Acc));
shrink([$  | T], [$  | Acc]) ->
  shrink(T, [$  | Acc]);
shrink([H | T], Acc) ->
  [H | shrink(T, Acc)].

flatten_test(X) ->
  shrink(X).

debug() ->
  dbg:start(),
  dbg:tracer(),
  dbg:tpl(?MODULE, '_', []),
  dbg:p(all, c).

debug2() ->
  dbg:tpl(?MODULE, '_', []).

%%phi() ->
%%  phi(1).
%%phi(X) ->
%%  Y = 1 + 1 / X,
%%  if X == Y -> X;
%%    true -> phi(Y)
%%  end.

File: js/src/bon.ts
import * as util from "util";
import {
    Atom,
    Binary,
    binary_to_list,
    byte_size,
    EmptyList,
    equal,
    float,
    int,
    integer_to_iolist,
    IOList,
    iolist_to_binary,
    list,
    List,
    list_to_array,
    list_to_atom,
    list_to_string,
    list_to_tuple,
    lists,
    map,
    string_to_binary,
    to_binary,
    Tuple,
    tuple_to_list,
    type_of
} from "./erlang-datatype";
import {$0, $a, $b, $colon, $double_quote, $quote, assert, char_code, test_out} from "./utils";

/*******
 * API *
 *******/

export function encode(Data): IOList {
    return serialize(Data);
}

export function decode(IOList_Bin) {
    const Type = type_of(IOList_Bin);
    switch (Type) {
        case "list":
            return decode(iolist_to_binary(IOList_Bin));
        case "binary":
            const List = binary_to_list(IOList_Bin);
            const [Rest, Res] = parse(List, EmptyList);
            if (Rest != EmptyList) {
                throw new TypeError("stream is not fully consumed: " + util.inspect({List, Rest, Res}));
            }
            return Res;
        default:
            throw new TypeError("invalid type: " + Type);
    }
}

export function decode_all(Data: IOList | Binary | List, Acc: List) {
    if (Data instanceof List && !Acc) {
        const IOList = Data as IOList;
        return decode(iolist_to_binary(IOList));
    } else if (Data instanceof Binary && !Acc) {
        const Bin = Data as Binary;
        const List = binary_to_list(Bin);
        return decode_all(List, EmptyList);
    } else if (Data instanceof List && Acc instanceof List) {
        const List_ = Data as List;
        const Case = parse(List_, EmptyList);
        if (Case.length == 2 && Case[0] == EmptyList) {
            const Res = Case[1];
            return lists.reverse(Acc.append(Res));
        }
        if (Case.length == 2) {
            const Next_Line = Case[0];
            const Res = Case[1];
            return decode_all(Next_Line, Acc.append(Res));
        }
        throw new TypeError("Failed to assign left hand side from right hand side: " + util.inspect(Case));
    } else {
        throw new Error("bad_arg");
    }
}

/********
 * Test *
 ********/

export function data_test(Data?) {
    if (Data) {
        const Res = decode(encode(Data));
        if (!equal(Data, Res)) {
            throw new Error("not_match: " + util.inspect({
                Data: {
                    type: type_of(Data),
                    value: Data
                }, Res: {
                    type: type_of(Res),
                    value: Res
                }
            }));
        }
        return "ok";
    }
    const res = [42,
        -72,
        1.5,
        3.14,
        Atom("atom"),
        "AB",
        [65, 0, 66],
        string_to_binary("text"),
        new Tuple(Atom("record"), Atom("content")),
        [new Tuple(Atom("debug"), Atom("true")), Atom("safe"), Atom("speedup"), new Tuple(Atom("log"), "file.log")],
        {user: "name", pw: "123"},
        {log: [{time: 123}, {time: 234}]}
    ].map(data => {
        const res = data_test(data);
        if (res == "ok") {
            test_out("passed " + util.inspect(data));
        }
        return res;
    }).every(res => res == "ok");
    return res === true ? "ok" : res;
}

/************
 * Internal *
 ************/

const WORD_TUPLE = "t";
const WORD_LIST = "l";
const WORD_MAP = "m";

/* serializer  */
export function serialize(Data): IOList {
    const type = type_of(Data);
    switch (type) {
        case "int":
            return list(32, integer_to_iolist(Data), 32);
        case "float": {
            const [A, B] = fac(Data);
            const Bin = B == 1 ? integer_to_iolist(A)
                : list(integer_to_iolist(A), char_code["/"], integer_to_iolist(B));
            return list(32, Bin, 32);
        }
        case "symbol": {
            const Bin = string_to_binary(Symbol.keyFor(Data));
            const Bin_Size = integer_to_iolist(byte_size(Bin));
            return list($quote, $a, $colon, Bin_Size, $colon, Bin, $quote);
        }
        case "string": {
            const Str = (Data as string).split('"').join('\\"');
            return list($double_quote, string_to_binary(Str), $double_quote);
        }
        case "array": {
            const Children = serialize_array(Data);
            return list(char_code["["], Children, 32, WORD_LIST, 32);
        }
        case "list": {
            const Children = serialize_list(Data);
            return list(char_code("["), Children, 32, WORD_LIST, 32);
        }
        case "binary": {
            const Bin = to_binary(Data);
            const Size = Bin.value.length;
            const Bin_Size = integer_to_iolist(Size);
            return list($quote, $b, $colon, Bin_Size, $colon, Bin, $quote);
        }
        case "tuple": {
            const Tuple = Data as Tuple;
            const List_ = tuple_to_list(Tuple);
            const Children = serialize_list(List_);
            return list(char_code["["], Children, 32, WORD_TUPLE, 32);
        }
        case "object": {
            const Children = Object.keys(Data)
                .reduce((Acc, K) => {
                    const V = Data[K];
                    const K_Bin = serialize(K);
                    const V_Bin = serialize(V);
                    return Acc.append(K_Bin).append(V_Bin);
                }, EmptyList);
            return list(char_code["["], Children, 32, WORD_MAP, 32);
        }
        default:
            throw new TypeError("unknown type: " + type + ", data=" + util.inspect(Data));
    }
}

function serialize_array(Array: any[]): IOList {
    return Array.reduce((Acc, X) => {
        return Acc.append(serialize(X));
    }, EmptyList);
}

function serialize_list(List: List): IOList {
    return lists.foldl((X, Acc) => Acc.append(serialize(X)), EmptyList, List);
}

/* parser */

interface Word {
    name: string
}

function parse(List: List, Acc: List): [List, any] {

    /* finish */
    if (List == EmptyList) {
        assert(Acc.tail == EmptyList, "invalid list");
        return [List, Acc.value];
    }

    const H = List.value;
    const T0 = List.tail;

    /* skip space */
    if (H == 32) {
        return parse(T0, Acc);
    }

    /* number */
    if (is_digit(H)) {
        const [num, tail] = parse_number(T0, H - 48, 1);
        return parse(tail, Acc.append(num));
    }
    if (H == char_code["-"] && T0 != EmptyList && is_digit(T0.value)) {
        const [num, tail] = parse_number(T0.tail, T0.value - 48, 1);
        return parse(tail, Acc.append(-num));
    }

    /* atom */
    {
        const [is_match, T1] = parse_head(list($quote, $a, $colon), List);
        if (is_match && is_digit(T1.value)) {
            const [Size, T2] = parse_number(T1.tail, T1.value - 48, 2);
            assert(T2.value == $colon, "invalid atom expr (1)");
            const [List, T3] = lists.split(Size, T2.tail);
            assert(T3.value == $quote, "invalid atom expr (2)");
            const Atom = list_to_atom(List);
            return parse(T3.tail, Acc.append(Atom));
        }
    }

    /* binary */
    {
        const match = () => {
            const H1 = H;
            if (H1 != $quote) {
                return [false];
            }
            const H2 = T0.value;
            if (H2 != $b) {
                return [false];
            }
            const H3 = T0.tail.value;
            if (H3 != $colon) {
                return [false];
            }
            const H4 = T0.tail.tail.value;
            if (!is_digit(H4)) {
                return [false];
            }
            const T0_ = T0.tail.tail.tail;
            return [true, H4, T0_];
        };
        const [is_match, H_, T0_] = match();
        if (is_match) {
            const [Size, List1] = parse_number(T0_, H_ - $0, 2);
            assert(List1.value == $colon, "Binary data should come with a colon after size");
            const T1 = List1.tail;
            const [List_, List2] = lists.split(Size, T1);
            assert(List2.value == $quote, "Binary data should end with a single quote");
            const T2 = List2.tail;
            const Bin = iolist_to_binary(List_);
            return parse(T2, Acc.append(Bin));
        }
    }

    /* group: tuple, list and map */
    if (H === char_code["["]) {
        const [Word, T1, Children] = parse(T0, EmptyList) as [any, any, any];
        const Res = (() => {
            switch (Word.name) {
                case WORD_TUPLE:
                    return list_to_tuple(Children);
                case WORD_LIST:
                    return list_to_array(Children);
                case WORD_MAP:
                    return list_to_map(Children, new Map());
                default:
                    throw new TypeError("unexpected word: " + util.inspect(Word));
            }
        })();
        return parse(T1, Acc.append(Res));
    }

    if (is_alphabet(H)) {
        const [Name, T1] = parse_word(T0, EmptyList.append(H));
        const Word: Word = {name: Name};
        return [Word, T1, Acc] as any;
    }

    /* string */
    if (H === $double_quote) {
        const [Str, T1] = parse_string(T0, "");
        return parse(T1, Acc.append(Str));
    }

    /* not impl */
    // List = list_to_string(List) as any;
    throw new Error("bad_arg: " + util.inspect({List, Acc}));
}

function parse_head(Token: List, List: List): [boolean, List] {
    if (Token == EmptyList) {
        return [true, List];
    }
    return Token.value == List.value
        ? parse_head(Token.tail, List.tail)
        : [false, List];
}

function parse_number(list: List, acc: number, count: number): [number, List] {
    if (list != EmptyList) {
        if (is_digit(list.value)) {
            return parse_number(list.tail, acc * 10 + (list.value - 48), count);
        }
        if (list.value == char_code["/"] && is_digit(list.tail.value) && count == 1) {
            const [q, tail] = parse_number(list.tail.tail, list.tail.value - 48, 2);
            return [acc / q, tail];
        }
    }
    return [acc, list];
}

/**
 * input without starting double quote
 * output without ending double quote
 *
 * e.g. 'text"rest' ~~> ["text","rest"]
 * */
function parse_string(list: List, acc: string): [string, List] {
    return list.value === $double_quote
        ? [acc, list.tail]
        : parse_string(list.tail, acc + String.fromCodePoint(list.value));
}

function parse_word(List_: List, Acc) {
    if (is_word_body(List_.value)) {
        return parse_word(List_.tail, Acc.append(List_.value));
    }
    assert(List_ instanceof List, "bad_arg first argument should be List");
    const Name = list_to_string(lists.reverse(Acc));
    return [Name, List_];
}

function fac(F: float) {
    if (F == 1) {
        return [1, 1];
    }
    if (F < 1) {
        const [A, B] = fac(1 / F);
        return fac2(B, A);
    }
    const [A, B] = fac_power_up(F, 1);
    return fac2(A, B);
}

export function fac_test(F) {
    const [A, B] = fac(F);
    const Diff = Math.abs(F - A / B);
    return Diff / F;
}

function fac2(A: int, B: int) {
    const D = gcd(A, B);
    return [round(A / D), round(B / D)];
}

const round = Math.round.bind(Math);

function fac_power_up(F: float, Acc: int) {
    return Math.round(F) == F ? [Math.round(F), Acc] : fac_power_up(F * 10, Acc * 10);
}

function gcd(A: int, B: int): int {
    return A == 0 ? B
        : B == 0 ? A
            : gcd(B, A % B);
}

function list_to_map(List: List, Acc: map) {
    if (List === EmptyList) {
        return Acc;
    }
    const K = List.value;
    const V = List.tail.value;
    const T = List.tail.tail;
    Acc.set(K, V);
    return list_to_map(T, Acc);
}

const is_digit = (c) => 48 <= c && c <= (48 + 9);
const is_small_cap = (c) => char_code["a"] <= c && c <= char_code["z"];
const is_large_cap = (c) => char_code["A"] <= c && c <= char_code["Z"];
const is_alphabet = (c) => is_small_cap(c) || is_large_cap(c);
const is_word_body = (c) => is_digit(c) || is_alphabet(c);
